File: main/find-the-nearest-clone/find-the-nearest-clone.cpp
#ifdef NDEBUG
#error Assertions unexpectedly disabled.
#endif

#include <algorithm>
#include <cassert>
#include <cstddef>
#include <iostream>
#include <iterator>
#include <queue>
#include <vector>

namespace {
    // An undirected graph, represented as an adjacency list.
    using Graph = std::vector<std::vector<int>>;

    // A visitation state for a vertex, tracking which root the traversal
    // started from (start) and its distance from that root (depth).
    // For a vertex that has not (yet?) been reached, start is zero.
    // This works because we are not also using 0 as a vertex label.
    struct State {
        int start;
        int depth;
    };

    // A visitation list for an undirected graph, tracking roots and distances.
    using Vis = std::vector<State>;

    Graph read_graph(const int vertex_count, int edge_count)
    {
        assert(vertex_count > 0 && edge_count > 0);

        Graph adj (vertex_count + 1);

        while (edge_count-- > 0) {
            int u {}, v {};
            std::cin >> u >> v;
            assert(0 < u && u <= vertex_count && 0 < v && v <= vertex_count);

            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        return adj;
    }

    template<typename Action>
    void read_roots_and_do(const int vertex_count, Action action)
    {
        assert(vertex_count >= 0);
        std::vector<int> colors (vertex_count + 1);
        std::copy_n(std::istream_iterator<int>{std::cin},
                    vertex_count, std::next(begin(colors)));

        int root_color {};
        std::cin >> root_color;
        assert(root_color > 0);

        for (auto vertex = 1; vertex <= vertex_count; ++vertex) {
            assert(colors[vertex] > 0);
            if (colors[vertex] == root_color) action(vertex);
        }
    }

    int multi_bfs(const Graph& adj, Vis& vis, std::queue<int>& queue)
    {
        while (!queue.empty()) {
            const auto src = queue.front();
            queue.pop();
            assert(vis[src].start > 0);

            for (const auto& dest : adj[src]) {
                if (vis[dest].start == vis[src].start) continue;

                if (vis[dest].start != 0)
                    return vis[src].depth + vis[dest].depth + 1;

                vis[dest].start = vis[src].start;
                vis[dest].depth = vis[src].depth + 1;
                queue.push(dest);
            }
        }

        return -1;
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    int vertex_count {}, edge_count {};
    std::cin >> vertex_count >> edge_count;

    const auto adj = read_graph(vertex_count, edge_count);
    Vis vis (vertex_count + 1);
    std::queue<int> queue;

    read_roots_and_do(vertex_count, [&vis, &queue](const int start) {
        vis[start].start = start;
        queue.push(start);
    });

    std::cout << multi_bfs(adj, vis, queue) << '\n';
}

File: main/find-the-nearest-clone/find-the-nearest-clone.py
#!/usr/bin/env python3

"""
"Find the nearest clone" - Finds the shortest distance in a graph between any
pair of vertices of a specified color.
"""

import collections


State = collections.namedtuple('State', ('start', 'depth'))
State.__doc__ = """
    A visitation state for a vertex, tracking which root the traversal started
    from (start) and its distance from that root (depth).
"""


def read_value():
    """Reads a line as a single integer."""
    return int(input())


def read_record(length):
    """Reads a line as a sequence of integers."""
    ret = list(map(int, input().split()))
    if len(ret) != length:
        raise ValueError('wrong record length')
    return ret


def read_graph(vertex_count, edge_count):
    """Reads an undirected graph as an adjacency list."""
    adj = [[] for _ in range(vertex_count + 1)]

    for _ in range(edge_count):
        u, v = read_record(2)
        adj[u].append(v)
        adj[v].append(u)

    return adj


def read_roots(vertex_count):
    """
    Reads a sequence of vertex colors followed by the root color, and yields
    the vertices whose colorings indicate they are roots.
    """
    colors = read_record(vertex_count)
    root_color = read_value()
    return (vertex for vertex, color in enumerate(colors, 1)
            if color == root_color)


def read_roots_as_vis_and_queue(vertex_count):
    """
    Reads a sequence of vertex colors followed by the root color, and returns
    an initialized visitation list and queue populated with the root vertices.
    """
    vis = [None for _ in range(vertex_count + 1)]
    queue = collections.deque()

    for root in read_roots(vertex_count):
        vis[root] = State(root, 0)
        queue.append(root)

    return vis, queue


def multi_bfs(adj, vis, queue):
    """Finds the shortest distance between root by multidirectional BFS."""
    while queue:
        src = queue.popleft()

        for dest in adj[src]:
            if vis[dest] is None:
                vis[dest] = State(vis[src].start, vis[src].depth + 1)
                queue.append(dest)
            elif vis[dest].start != vis[src].start:
                return vis[src].depth + vis[dest].depth + 1

    return None


def run():
    """Solves the problem described by values read from standard input."""
    vertex_count, edge_count = read_record(2)
    adj = read_graph(vertex_count, edge_count)
    vis, queue = read_roots_as_vis_and_queue(vertex_count)
    result = multi_bfs(adj, vis, queue)
    print(-1 if result is None else result)


if __name__ == '__main__':
    run()
